---
title: "Assignment 3"

author: "Dylan Renaud, Srikrishnan Ramachandran,Fezan Mukhtar"

date: '2022-05-29'
output: html_document
---

```{r, Question 1}
# Question 1. Read in the file, making the gene identifiers the row names. Show a table of values for the first six genes

URL<- "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv"

download.file(URL,destfile="gene_expression")

read.table(URL)


data<-read.table(URL)
data<-read.table(URL,stringsAsFactors=FALSE, header=TRUE)

data<-read.table(URL,stringsAsFactors=FALSE, header=TRUE, row.names=1)
head(data)
str(data)
rownames(data)
grep("ENG",rownames(data))
data[1:6,]

```

```{r,Question2}
#Question 2: Make a new column which is the mean of the other columns. Show a table of values for the first six genes.
ncol(data)
rowMeans(data)
Mean<-rowMeans(data)
Mean
data$Mean<-Mean
head(data)
#First six genes
data[1:6,]

```





```{r,Question3}
# Question 3: list 10 genes with the highest mean expression.
order(data)
order(data$Mean)
data$Mean
data[order(data$Mean),]
data[order(-data$Mean),]
revsort<-data[order(-data$Mean),]
revsort[1:10,c(4,ncol(revsort))]
revsort[1:10,]





```


```{r, Question 4}
#Question 4:Determine the number of genes with a mean <10
data$Mean<10
which(data$Mean<10)
data[which(data$Mean<10),]
dt<-data[which(data$Mean<10),]
nrow(dt)

```

```{r, Question5}
# Question 5 Make a histogram plot of the mean values and include it into your report.
as.matrix(data$Mean)
dmat<-as.matrix(data$Mean)
dmat
hist(dmat,xlab = "Mean values",main = "Mean of gene expressions", breaks=100)


```




```{r, Question 6}
url="https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv"
read.table(url,sep = ",",header=TRUE,)
inf<-read.table(url,sep = ",",header=TRUE,)
head(inf)
colnames(inf)
```

```{r, Question7}
#Question 7: Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.
meanstart<-mean(inf$Circumf_2005_cm)
meanstart
sdstart<-sd(inf$Circumf_2005_cm)
sdstart
meanend<-mean(inf$Circumf_2020_cm)
meanend
sdend<-sd(inf$Circumf_2020_cm)
sdend

```





```{r, question 8}
#Question 8:Make a box plot of tree circumferance at the start and end of the study of both sites.

site<-inf$Site
site
rownames(inf)
colnames(inf)
ne<-grep("northeast",inf$Site)
ne
sw<-grep("southwest",inf$Site)
sw
ne2020<-inf[ne,6]
ne2020
ne2005<-inf[ne,3]
ne2005
sw2020<-inf[sw,6]
sw2020
sw2005<-inf[sw,3]
sw2005
boxplot(ne2005,ne2020,sw2005,sw2020, ylab="Tree circumference",main="Growth measurements", at=c(1,2,3,4),
        names =c("northeast 2005","northeast 2020","southwest 2005","southwest 2020"))
```

```{r, Question9}

 mean_2010 <- mean(inf$Circumf_2010_cm)
 mean_2010
 
 mean_2020 <- mean(inf$Circumf_2020_cm)
 mean_2020
 
 Growth_rate <- mean_2020 - mean_2010 
 Growth_rate


```

```{r, Question10}
#Question 10
head(t.test(meangrowthsw,meangrowthne))
diff1<-ne2020-ne2010
diff2<-sw2020-sw2010
rest.test<-t.test(diff1, diff2)
pval <- rest.test$p.value
pval
rest.test2<- wilcox.test(diff1,diff2)
pval2<-rest.test2$p.value
pval2
```


```{r, Question2(Part 2)}
#Question 2:
#E. coli

head(summary(cds))
str(summary(cds))
head(summary(cds))[,1]
ecol<-as.numeric(summary(cds)[,1])
sum(ecol)

#Anaerococcus
head(summary(an))
str(summary(an))
head(summary(an))[,1]
anna<-as.numeric(summary(an)[,1])
summary(an)
sum(anna)
#sum=2011938
boxplot(anna,ecol)
boxplot((anna),(ecol), ylab="Sequence length",
        main="Comparison of sequnce length between E. coli and anaerococcus",
        names =c( "anaerococcus","E. coli"))

```



```{r,(2) Question 4}
#Question 4:Create a codon usage table and quantify the codon usage bias among all coding sequences. Describe
#any differences between the two organisms with respect to their codon usage bias. Provide charts to
#support your observations.


#E coli
uco(cds[[2]],index="rscu",as.data.frame=TRUE)
ecodon<-uco(cds[[2]],index="rscu",as.data.frame=TRUE)

ecodon %>%
  kbl() %>%
  kable_paper("hover",full_width=F)

lowbias<-ecodon[order(ecodon$freq),]

lowbias%>%
  kbl()%>%
  kable_paper("hover",full_width=F)  

highb<-ecodon[order(-ecodon$freq),]

highb%>%
  kbl()%>%
  kable_paper("hover",full_width=F)  
 
#anaerococcus
uco(an[[2]],index="rscu",as.data.frame=TRUE)
acodon<-uco(an[[2]],index="rscu",as.data.frame=TRUE)

acodon %>%
  kbl() %>%
  kable_paper("hover",full_width=F)

alowbias<-acodon[order(acodon$freq),]


alowbias%>%
  kbl()%>%
  kable_paper("hover",full_width=F)  

higha<-acodon[order(-acodon$freq),]


higha%>%
  kbl()%>%
  kable_paper("hover",full_width=F)

hist(acodon$freq,xlab = "Frequency",ylab="Number of codons", main= "anaerococcus")
hist(ecodon$freq,xlab = "Frequency",ylab="Number of codons",main= "E coli")
```





```{r, Question 5}
#Question 5: In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most over and under-represented k-mers in your organism of interest. Are these k-mers also over- and
#under-represented in E. coli to a similar extent? Provide plots to support your observations. Why do
#you think these sequences are present at different levels in the genomes of these organisms?

pro<-unlist(prot)
anpro<-unlist(anprot)
count(pro,alphabet= aa,wordsize=3, freq=TRUE)

#3

#Anaerococcus
acomp3<-count(anpro,alphabet= aa,wordsize=3, freq=TRUE)
aoverep3<-head(sort(acomp3,decreasing=TRUE),n=10L)
aoverep3
aunderep3<-head(sort(acomp3,increasing=TRUE),n=10L)
aunderep3

#E coli
ecomp3<-count(pro,alphabet= aa,wordsize=3, freq=TRUE)

grep(("LLL"),rownames(ecomp3))
grep(("FFL"),rownames(ecomp3))
grep(("LSL"),rownames(ecomp3))
grep(("SLL"),rownames(ecomp3))
grep(("SSS"),rownames(ecomp3))
grep(("LFL"),rownames(ecomp3))
grep(("SSL"),rownames(ecomp3))
grep(("SLS"),rownames(ecomp3))
grep(("LIL"),rownames(ecomp3))
grep(("LLI"),rownames(ecomp3))
eoverep3<-ecomp3[c(3790,1790,3910,6190,6316,3690,6310,6196,3750,3788)]
eoverep3

grep(("CDW"),rownames(ecomp3))
grep(("HEW"),rownames(ecomp3))
grep(("HMC"),rownames(ecomp3))
grep(("HWG"),rownames(ecomp3))
grep(("MCM"),rownames(ecomp3))
grep(("MHW"),rownames(ecomp3))
grep(("VMC"),rownames(ecomp3))
grep(("WDP"),rownames(ecomp3))
grep(("WHM"),rownames(ecomp3))
grep(("WMC"),rownames(ecomp3))
eunderep3<-ecomp3[c(459,2479,2602,2766,4031,4139,7002,7253,7331,7402)]
eunderep3



#4

#anaerococcus
acomp4<-count(anpro,alphabet= aa,wordsize=4, freq=TRUE)
acomp4
aoverep4<-head(sort(acomp4,decreasing=TRUE),n=10L)
aoverep4
aunderep4<-head(sort(acomp4,increasing=TRUE),n=10L)
aunderep4



ecomp4<-count(pro,alphabet= aa,wordsize=4, freq=TRUE)

grep(("LLLL"),rownames(ecomp4))
grep(("LLFL"),rownames(ecomp4))
grep(("FLLL"),rownames(ecomp4))
grep(("LFLL"),rownames(ecomp4))
grep(("LLSL"),rownames(ecomp4))
grep(("LLIL"),rownames(ecomp4))
grep(("LSLS"),rownames(ecomp4))
grep(("SSSS"),rownames(ecomp4))
grep(("SLSL"),rownames(ecomp4))
grep(("SLLL"),rownames(ecomp4))
eoverep4<-ecomp4[c(75790,75690,35790,73790,75910,75750,78196,126316,123910,123790)]
eoverep4

grep(("AAAC"),rownames(ecomp4))
grep(("AAAH"),rownames(ecomp4))
grep(("AAAP"),rownames(ecomp4))
grep(("AAAV"),rownames(ecomp4))
grep(("AAAY"),rownames(ecomp4))
grep(("AACC"),rownames(ecomp4))
grep(("AACD"),rownames(ecomp4))
grep(("AACE"),rownames(ecomp4))
grep(("AACH"),rownames(ecomp4))
grep(("AACK"),rownames(ecomp4))
eunderep4<-ecomp4[c(2,7,13,18,20,22,23,24,27,29)]
eunderep4



#5

#anaerococcus
acomp5<-count(anpro,alphabet= aa,wordsize=5, freq=TRUE)
acomp5
aoverep5<-head(sort(acomp5,decreasing=TRUE),n=10L)
aoverep5
aunderep5<-head(sort(acomp5,increasing=TRUE),n=10L)
aunderep5

ecomp5<-count(pro,alphabet= aa,wordsize=5,freq=TRUE)
ecomp5

grep(("LLLLL"),rownames(ecomp5))
grep(("FLLLL"),rownames(ecomp5))
grep(("LLLFL"),rownames(ecomp5))
grep(("LLFLL"),rownames(ecomp5))
grep(("LLLIL"),rownames(ecomp5))
grep(("SLSLS"),rownames(ecomp5))
grep(("LFLLL"),rownames(ecomp5))
grep(("LSFLL"),rownames(ecomp5))
grep(("LSSSS"),rownames(ecomp5))
grep(("SISSS"),rownames(ecomp5))
eoverep5<-ecomp5[c(1515790,715790,1515690,1513790,1515750,2478196,1475790,1561790,1566316,2462316)]
eoverep5

grep(("AAAAC"),rownames(ecomp5))
grep(("AAAAD"),rownames(ecomp5))
grep(("AAAAF"),rownames(ecomp5))
grep(("AAAAH"),rownames(ecomp5))
grep(("AAAAK"),rownames(ecomp5))
grep(("AAAAL"),rownames(ecomp5))
grep(("AAAAN"),rownames(ecomp5))
grep(("AAAAP"),rownames(ecomp5))
grep(("AAAAR"),rownames(ecomp5))
grep(("AAAAS"),rownames(ecomp5))
eunderep5<-ecomp5[c(2,3,5,7,9,10,12,13,15,16)]
eunderep5

#Plots

#E coli
plot(eoverep3,type="b", col="yellow", xlab="codon",ylab="protein frequency",main="E coli")
lines(eoverep4,type="b", col="pink")
lines(eoverep5,type="b", col="gray")
lines(eunderep3,type="b", col="blue")
lines(eunderep4,type="b", col="red")
lines(eunderep5,type="b", col="black")


#Anaerococcus
plot(aoverep3,type="b", col="yellow", xlab="codon",ylab="protein frequency", main="Anaerococcus")
lines(aoverep4,type="b", col="pink")
lines(aoverep5,type="b", col="gray")
lines(aunderep3,type="b", col="blue")
lines(aunderep4,type="b", col="red")
lines(aunderep5,type="b", col="black")

```